app/server/lib/AttachmentFileManager.ts

```diff
@@ -108,24 +144,253 @@ export class AttachmentFileManager implements IAttachmentFileManager {
+  // transfers them from their current store to their target store. This ensures that for a given
+  // doc, we never accidentally start several transfers at once and load many files into memory
+  // simultaneously (e.g. a badly written script spamming API calls). It allows any new transfers
+  // to overwrite any scheduled transfers. This provides a well-defined behaviour where the latest
+  // scheduled transfer happens, instead of the last transfer to finish "winning".
+  public startTransferringFileToOtherStore(fileIdent: string, newStoreId: AttachmentStoreId | undefined) {
+    this._pendingFileTransfers.set(fileIdent, newStoreId);
+    this._runTransferJob();
+  }
+
+  // Generally avoid calling this directly, instead use other methods to schedule and run the
+  // transfer job. If a file with a matching identifier already exists in the new store, no
+  // transfer will happen, as the default _addFileToX behaviour is to avoid re-uploading files.
+  public async transferFileToOtherStore(fileIdent: string, newStoreId: AttachmentStoreId | undefined): Promise<void> {
+    this._log.info({fileIdent, storeId: newStoreId}, `transferring file to new store`);
+    const fileMetadata = await this._docStorage.getFileInfoNoData(fileIdent);
+    // This check runs before the file is retrieved as an optimisation to avoid loading files into
+    // memory unnecessarily.
+    if (!fileMetadata || fileMetadata.storageId == newStoreId) {
+      return;
+    }
+    // It's possible that the record has changed between the original metadata check and here.
+    // However, the worst case is we transfer a file that's already been transferred, so no need to
+    // re-check.
+    // Streaming isn't an option here, as SQLite only supports buffers (meaning we need to keep at
+    // least 1 full copy of the file in memory during transfers).
+    const file = await this._getFile(fileIdent);
+    if (!await validateFileChecksum(fileIdent, file.data)) {
+      throw new AttachmentRetrievalError(file.storageId, file.ident, "checksum verification failed for retrieved file");
+    }
+    if (!newStoreId) {
+      await this._storeFileInLocalStorage(fileIdent, file.data);
+      return;
+    }
+    const newStore = await this._getStore(newStoreId);
+    if (!newStore) {
+      this._log.warn({
+        fileIdent,
+        storeId: newStoreId
+      }, `unable to transfer file to unavailable store`);
+      throw new StoreNotAvailableError(newStoreId);
+    }
+    // Store should error if the upload fails in any way.
+    await this._storeFileInAttachmentStore(newStore, fileIdent, file.data);
+
+    // Don't remove the file from the previous store, in case we need to roll back to an earlier
+    // snapshot.
+    // Internal storage is the exception (and is automatically erased), as that's included in
+    // snapshots.
+  }
+
+  public allTransfersCompleted(): Promise<void> {
```

Comment:
Is this only used in tests? If only used in tests, codebase convention is to prefix name with test, e.g. `testAllTransfersCompleted` and maybe mention that fact in a comment.

The name is perhaps a little misleading in the case of there never being any transfers, but no big deal if just for tests anyway.

Reply:
Code would maybe be simpler to read if you marked the method as `async`, wouldn't need that Promise.resolve()

Could be just `await this._transferJob?.()`